### PR TITLE
"registry upload artifact" subcommand

### PR DESCRIPTION
**Pull request by timburks**
This adds a general subcommand to "registry upload" that uploads artifacts of a variety of types, using the "Kind" field in the artifact to identify the type. Currently manifests, taxonomy lists, and lifecycles are supported, but ideally all known artifacts would be supported by this command.

Addresses #413

TODO (in this or future PRs):
- tests for uploading lifecycles and taxonomy lists.
- include other types, such as style guides.
- deprecate "registry upload manifest" and "registry upload styleguide".

### CODE REVIEW

**Comment by timburks** [comment]
TODO: tests for uploading lifecycles and taxonomy lists.
TODO: include other types, such as style guides.
TODO: deprecate "registry upload manifest" and "registry upload styleguide".

**Review by shrutiparabgoogle — APPROVED** [review]

File: cmd/registry/cmd/upload/artifact.go [hunk 1]
```diff
@@ -0,0 +1,166 @@
...
+
+func artifactCommand(ctx context.Context) *cobra.Command {
+	var parent string
+	cmd := &cobra.Command{
+		Use:   "artifact FILE_PATH --parent=value",
+		Short: "Upload an artifact",
+		Args:  cobra.MinimumNArgs(1),
+		Run: func(cmd *cobra.Command, args []string) {
+			artifactPath := args[0]
+			if artifactPath == "" {
```
Comment (shrutiparabgoogle) [inline cmd/registry/cmd/upload/artifact.go:41]: NIT: Suggest to rename this as `artifactFilePath`, my first interpretation of `artifactPath` was path in the registry like  `projects/demo/locations/global/artifacts/manifest`

File: cmd/registry/cmd/upload/artifact_test.go [hunk 2]
```diff
@@ -0,0 +1,118 @@
...
+	"github.com/apigee/registry/rpc"
+	"github.com/google/go-cmp/cmp"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+	"google.golang.org/protobuf/proto"
+	"google.golang.org/protobuf/testing/protocmp"
+)
+
+func TestManifestArtifactUpload(t *testing.T) {
+	tests := []struct {
```
Comment (shrutiparabgoogle) [inline cmd/registry/cmd/upload/artifact_test.go:33]: Looks like these tests are missing some error path testing. I can take it up as a TODO to add more tests here in manifest upload

File: cmd/registry/cmd/upload/artifact_test.go [hunk 3]
```diff
@@ -0,0 +1,118 @@
+// Copyright 2021 Google LLC
```
Comment (seaneganx) [inline cmd/registry/cmd/upload/artifact_test.go:1]: 2022

File: cmd/registry/cmd/upload/artifact.go [hunk 4]
```diff
@@ -0,0 +1,166 @@
...
+	if err != nil {
+		return nil, err
+	}
+	return &rpc.Artifact{
+		Contents: artifactBytes,
+		MimeType: core.MimeTypeForMessageType("google.cloud.apigeeregistry.v1.controller.Manifest"),
+	}, nil
+}
+
+func validateManifest(ctx context.Context, m *rpc.Manifest) error {
```
Comment (seaneganx) [inline cmd/registry/cmd/upload/artifact.go:122]: I think this belongs in the controller package. Does it have validation code that can be exported and reused, perhaps after some minor refactoring?

**Review by seaneganx — APPROVED** [review]

File: cmd/registry/cmd/upload/artifact_test.go [hunk 5]
```diff
@@ -0,0 +1,118 @@
...
+			_, err = adminClient.CreateProject(ctx, &rpc.CreateProjectRequest{
+				ProjectId: test.project,
+				Project: &rpc.Project{
+					DisplayName: "Demo",
+					Description: "A demo catalog",
+				},
+			})
+			if err != nil {
+				t.Fatalf("Failed to create project %s: %s", test.project, err.Error())
+			}
```
[the remark this answers is not in the document]
Comment (shrutiparabgoogle) [reply cmd/registry/cmd/upload/artifact_test.go:90]: These are tests specific to `manifest upload`. I can look at refactoring this.

File: cmd/registry/cmd/upload/artifact.go (hunk 4 above)
Comment (shrutiparabgoogle) [reply cmd/registry/cmd/upload/artifact.go:122]: I can look at this while adding error tests for manifest upload.

File: cmd/registry/cmd/upload/artifact.go (hunk 1 above)
Comment (timburks) [reply cmd/registry/cmd/upload/artifact.go]: Thanks - I've changed the name

File: cmd/registry/cmd/upload/artifact.go [hunk 6]
```diff
@@ -0,0 +1,166 @@
...
+)
+
+func artifactCommand(ctx context.Context) *cobra.Command {
+	var parent string
+	cmd := &cobra.Command{
+		Use:   "artifact FILE_PATH --parent=value",
+		Short: "Upload an artifact",
+		Args:  cobra.MinimumNArgs(1),
+		Run: func(cmd *cobra.Command, args []string) {
+			artifactPath := args[0]
```
[the remark this answers is not in the document]
Comment (timburks) [reply cmd/registry/cmd/upload/artifact.go]: Thanks, I was unaware of ExactArgs() - this one is changed.

File: cmd/registry/cmd/upload/artifact_test.go [hunk 7]
```diff
@@ -0,0 +1,118 @@
...
+			if err = cmd.Execute(); err != nil {
+				t.Fatalf("Execute() with args %v returned error: %s", args, err)
+			}
+
+			req := &rpc.GetArtifactContentsRequest{
+				Name: "projects/" + test.project + "/locations/global/artifacts/test-manifest",
+			}
+
+			manifest := &rpc.Manifest{}
+			body, _ := client.GetArtifactContents(ctx, req)
```
[the remark this answers is not in the document]
Comment (timburks) [reply cmd/registry/cmd/upload/artifact_test.go]: Fixed.

File: cmd/registry/cmd/upload/artifact.go [hunk 8]
```diff
@@ -0,0 +1,166 @@
...
+			}
+			artifact, artifactID, err := buildArtifact(ctx, artifactPath)
+			if err != nil {
+				log.FromContext(ctx).WithError(err).Fatal("Failed to read artifact")
+			}
+			client, err := connection.NewClient(ctx)
+			if err != nil {
+				log.FromContext(ctx).WithError(err).Fatal("Failed to get client")
+			}
+			artifact.Name = fmt.Sprintf("%s/artifacts/%s", parent, artifactID)
```
[the remark this answers is not in the document]
Comment (timburks) [reply cmd/registry/cmd/upload/artifact.go]: I moved this into the top-level buildArtifact() function - pushing it lower seemed repetitive, but other suggestions are welcome :)

File: cmd/registry/cmd/upload/artifact_test.go (hunk 3 above)
Comment (timburks) [reply cmd/registry/cmd/upload/artifact_test.go]: 🤦 thanks

File: cmd/registry/cmd/upload/artifact.go (hunk 4 above)
Comment (timburks) [reply cmd/registry/cmd/upload/artifact.go:123]: SGTM - thanks @shrutiparabgoogle

File: cmd/registry/cmd/upload/artifact.go [hunk 9]
```diff
@@ -0,0 +1,166 @@
+// Copyright 2021 Google LLC
```
[the remark this answers is not in the document]
Comment (timburks) [reply cmd/registry/cmd/upload/artifact.go]: ack

**Pull request merged**